### PR TITLE
:seedling: Add support for CA/certificate rotation

### LINKED ISSUE #915: Handle CA rotation for Catalogd web server trust
With the addition of TLS communication between operator-controller and catalogd, [we now mount](https://github.com/operator-framework/operator-controller/pull/888) the CA certificate into operator-controller so it can trust connections to the catalogd web server.

It would be ideal if we could handle rotations of the mounted CA certificate without restarting the manager. As of now, the contents of the mounted CA certificate *do* change on disk when the certificate rotates, but the manager does not pick up any changes since Go [does not have a mechanism](https://github.com/golang/go/issues/35887#issuecomment-1798581909) for reloading rootCA.

There is the possibility to implement custom connection verification logic in [tls.Config.VerifyConnection()](https://pkg.go.dev/crypto/tls#example-Config-VerifyConnection), but for this to work you would also need to set InsecureSkipVerify to true which isn't ideal. The ideal solution would probably be to contribute something akin to GetCertificate but for the rootCA upstream into Go and then use that feature in operator-controller.

It might also be worth modifying the CertificateWatcher in controller-runtime so it can function without providing a key--currently it requires both the cert and key since it's built for managing rotation on servers. A GetRootCA-type functionality for re-loading the CA cert in tls.Config would utilize the modified CertificateWatcher in a similar manner to how the certificate rotation is [handled in Catalogd](https://github.com/operator-framework/catalogd/blob/main/internal/serverutil/serverutil.go#L33).

### PR DESCRIPTION
**Pull request by tmshort**
Fixes #915
Mounted secrets are automatically updated into pods, but...
* It doesn't work with `subPath` mountings
* When `subPath` is not used, then a bunch of directories are mounted
* And one of those directories is a symlink, so `IsDir()` returns false
* And a watch is needed to notice the change

So, update the certificate volume patch, which requires a change in how we look for certificates in the CA cert directory.

Add a watch, so when the certs do change, we update the cert pool.
Also look at validity dates of certificates, and error on expired certs.

The default cert-manager certificates have 90 days validities.

# Description

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

### CODE REVIEW

**Comment by tmshort** [comment]
This also adds a unit-test for an empty cert dir, and for an expired certificate. And also adds an e2e to make sure secret updates and rotations actually work.

File: cmd/manager/main.go
Comment (everettraven) [inline cmd/manager/main.go:181]: I'm a bit concerned with this approach from the perspective of a user that might see the pod having restarts and suspecting it is a bug, filing a bug report, and then us as maintainers having to perform deeper investigation to find that it was a cert rotation and stating that it is expected behavior for the pod to "crash" to reload CA certificates.

While it may be easier to simply crash the pod and it be restarted, I think it will throw a false negative signal to users. I think a caching based approach, while more complex, would ensure we aren't issuing false negative signals.

Comment (tmshort) [reply cmd/manager/main.go:181]: It's clear from the (prior) logs that it's a restart from cert rotation.
CA certificates should rarely, if ever, rotate. We're talking a minimum of 3 months (for default cert-manager certificates) to years for commercial CAs.
According to Steve K, this is a common approach (I specifically asked him about it in another project).

Not sure what you mean by a cache-based approach? These are files mounted in the pod itself, not Secrets or ConfigMaps that are being monitored.

File: cmd/manager/main.go
Comment (everettraven) [inline cmd/manager/main.go:279]: Instead of updating the cert pool on the ClusterExtensionReconciler, what if we instead used a type that cached this information and that object was passed around?

So instead of having to do something like:
```go
	ceRecon := &controllers.ClusterExtensionReconciler{}

	unpacker := &source.ImageRegistry{
		BaseCachePath: filepath.Join(cachePath, "unpack"),
		// TODO: This needs to be derived per extension via ext.Spec.InstallNamespace
		AuthNamespace: systemNamespace,
		GetCaCertPool: func() (*x509.CertPool, error) {
			ceRecon.CaMutex.RLock()
			defer ceRecon.CaMutex.RUnlock()
			return ceRecon.CaCertPool.Clone(), nil
		},
	}
```

it could instead be:
```go
        certPooler := httputil.NewCertPooler(...)
	unpacker := &source.ImageRegistry{
		BaseCachePath: filepath.Join(cachePath, "unpack"),
		// TODO: This needs to be derived per extension via ext.Spec.InstallNamespace
		AuthNamespace: systemNamespace,
		// Used the existing approach as a way to show the example, but this could
		// probably also just be updated to be a cert pooler instead of a function
		// although keeping it as a function is also fine by me
		GetCaCertPool: func() (*x509.CertPool, error) {
			return certPooler.Get()
		},
	}
```

Comment (tmshort) [reply cmd/manager/main.go:279]: Hmmm.... Would still need to store the pool somewhere, and that place right now is on the reconciler, but doesn't have to be. I can look at this,. but not sure how much "tidier" it would be.

Comment (everettraven) [reply cmd/manager/main.go:279]: The pool would be stored on the `CertPoolWatcher`(?) type

File: hack/test/cert-e2e.sh
Comment (m1kola) [inline hack/test/cert-e2e.sh:31]: This is quite hard to follow, so I might be missing something. But it seems to me that we will never exit in the unhappy path (string never appears in logs) because we are waiting for `NEXT_WAIT_TIME` to be equal to `MAX_TIME` and `NEXT_WAIT_TIME` never defined.

Comment (m1kola) [reply hack/test/cert-e2e.sh:31]: If you decide to switch to Go tests for this at some point - you might find this useful:

https://github.com/operator-framework/operator-controller/blob/95b9f0dc5943497028deb47969310dec45792a2d/test/upgrade-e2e/post_upgrade_test.go#L116-L145

File: internal/catalogmetadata/cache/cache.go
Comment (m1kola) [inline internal/catalogmetadata/cache/cache.go:101]: As far as I understand this PR - we are creating a new client each time we need to make a request to catalogd to make sure that we read a new cert.

But if it is the case - why do we need a cert watcher?

**Review by m1kola — COMMENTED** [review]
I'm still reviewing.

For these interested, here are some relevant issues in Go and K8s on root CA reloading:

* [Go lang proposal for root CA reloading](https://github.com/golang/go/issues/64796)
* [K8s client-go issue for root CA reloading](https://github.com/kubernetes/kubernetes/issues/119483)

TLDR: it looks like there is no equivalent of `GetClientCertificate`/`GetCertificate`, but for root CAs in TLS config in standard library, but it is possible to implement a workaround with `VerifyPeerCertificate`. This would require setting `InsecureSkipVerify` to `true` and re-implementing the standard verification in `VerifyPeerCertificate` as far as I understand - see [this example](https://pkg.go.dev/crypto/tls@master#example-Config-VerifyConnection) from Go docs.

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh:31]: As of now, it doesn't fail... but you are correct. I didn't update every variable.

**Comment by tmshort** [comment]
> TLDR: it looks like there is no equivalent of `GetClientCertificate`/`GetCertificate`, but for root CAs in TLS config in standard library, but it is possible to implement a workaround with `VerifyPeerCertificate`. This would require setting `InsecureSkipVerify` to `true` and re-implementing the standard verification in `VerifyPeerCertificate` as far as I understand - see [this example](https://pkg.go.dev/crypto/tls@master#example-Config-VerifyConnection) from Go docs.


And we really don't want to do that (set `InsecureSkipVerify`) by default. And reimplementing the verification code is probably not worth it, when we can just offer up the updated certificate pool on demand.

File: internal/catalogmetadata/cache/cache.go
Comment (tmshort) [reply internal/catalogmetadata/cache/cache.go:101]: Because we don't want to read the certs off disk every single time we want to make a connection. The cert pool is used in multiple places.

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh:21]: You're correct in that the cert-e2e only test the reloading of certificates. It would be _much_ more complex and time-consuming to test an end-to-end connection with catalogd. The renewal period would have to be adjusted to make this test possibly run 10+ minutes, which is undesirable. Now that there's a separate CertPoolWatcher, I can likely do it as a unit test; the code before was not amenable to a unit-test.

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh:17]: You might be right, I can verify. But if I can use a unit test instead, this wouldn't be necessary.

File: internal/httputil/certpoolwatcher.go
Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:78]: I can break it up (specifically the update function), which would make it a bit easier to understand.

Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:78]: The complexity is there due to the number if `func()` that are present due to the use of `defer` and `return`...

File: internal/catalogmetadata/cache/cache.go
Comment (m1kola) [reply internal/catalogmetadata/cache/cache.go:101]: Makes sense. Resolving this question.

**Comment by m1kola** [comment]
> And we really don't want to do that (set `InsecureSkipVerify`) by default. And reimplementing the verification code is probably not worth it, when we can just offer up the updated certificate pool on demand.

Setting `InsecureSkipVerify` to `true` and `VerifyPeerCertificate` means that we take over the responsibility of verifying the certs from standard library. That's not great, but it is not a lot of code (see standard library [code](https://github.com/golang/go/blob/70491a81113e7003e314451f3e3cf134c4d41dd7/src/crypto/tls/handshake_client.go#L1117-L1128), and [client example](https://pkg.go.dev/crypto/tls@master#example-Config-VerifyConnection) from the docs). It doesn't sounds too terrible to me, so I'm open to consider this.

As far as I understand in this implementation we are re-creating a client each time and will have to establish a connection each we use it which is not optimal.

I don't know how significantly it will affect our performance. One one hand - creating new connections every time is not great, but on the other hand - not sure if we are going to benefit from re-using the connection in these use cases.

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh]: I removed the e2e, and now use a unit test

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh]: I removed the e2e and now use a unit test

File: internal/httputil/certpoolwatcher.go
Comment (tmshort) [reply internal/httputil/certpoolwatcher.go]: I broke it up, and @joelanford offered some suggests for performance, so there's still a `drain()` function, but it's isolated.

File: hack/test/cert-e2e.sh
Comment (tmshort) [reply hack/test/cert-e2e.sh]: I removed the e2e in favor of a unit test

File: Makefile
Comment (tmshort) [reply Makefile]: I removed the e2e in favor of a unit test

File: internal/httputil/certpoolwatcher.go
Comment (bentito) [inline internal/httputil/certpoolwatcher.go:62]: Could there be spurious events in this dir that are going to cause us to do work when we don't really need to? Just wondering how clean cert manager, and anything else with access, is in terms of usage of this dir? Maybe we should filter it down further or just watch .crt files, etc.

Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:62]: That's what this function is for... it's used to "drain" the event cue to avoid spurious events.
It's really just kubernetes that has write access, it updates the certs based on when a Secret is updated. The directory is read-only within the pod.

Comment (bentito) [reply internal/httputil/certpoolwatcher.go:62]: sounds like it'll be okay then

File: internal/httputil/certpoolwatcher.go
Comment (everettraven) [inline internal/httputil/certpoolwatcher.go:71]: What happens with the watcher in the case that the close fails? I see the goroutine for handling events exits, but does the watcher continue to stick around filling the event queue and consuming memory?

File: internal/httputil/certpoolwatcher.go
Comment (everettraven) [inline internal/httputil/certpoolwatcher.go:85]: Is it expected that encountering an error when performing an update to the certificate pool that restarting the pod will result in the error being resolved?

File: internal/httputil/certpoolwatcher.go
Comment (everettraven) [inline internal/httputil/certpoolwatcher.go:63]: Do you know if any events include _only_ the directory being watched as the event.Name value? If so, I wonder if instead of performing this drainEvents action we could do some event filtering similar to https://github.com/fsnotify/fsnotify/blob/c1467c02fba575afdb5f4201072ab8403bbf00f4/cmd/fsnotify/file.go#L66-L78

I won't block the PR merging on this, but something that could make it so we don't have any "sleep" actions

Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:63]: Some filtering might be useful. The only time this path should be updated is when a Secret is updated. The directory is read-only within the pod.

File: internal/httputil/certpoolwatcher.go
Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:85]: If we can't create the cert pool, then something is wrong with the pod. The mount of the certs directory is broken, or the set of certificates is bad, or the configuration of the pod is bad. Joe has given a previous comment he'd rather have the pod crash (and be noticed) than have it continue to run in an error condition.

File: internal/httputil/certpoolwatcher.go
Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:71]: Since the `cpw` is still likely referenced _somewhere_ the memory for the notifier is still around; but the notify OS resources have been cleaned up the _best that they can be_.

File: internal/httputil/certpoolwatcher.go
Comment (everettraven) [reply internal/httputil/certpoolwatcher.go:63]: >We need to recognize new files, deleted files, updated files, etc.
If we only react to "directory has been updated" type events wouldn't we catch these events as well?

**Review by everettraven — APPROVED** [review]

File: internal/httputil/certpoolwatcher.go
Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:63]: But it wouldn't catch updates to files within, as that's a change to the file, not the directory.

Comment (everettraven) [reply internal/httputil/certpoolwatcher.go:63]: I thought the reasoning for the drain events operation was because when we receive updates we get mass events on everything when something changed. Maybe I misunderstood, which led me to thinking that if any change happened in the directory (including an individual file), it would trigger an event for the directory as well.

Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:63]: The Watch is on the directory, and that includes the contents. It also depends on how things are mounted. A change to a file within a directory does not necessarily indicate a change to the directory.
The drain is there because the update of a single secret may trigger a number of events (I was seeing 4+, because of how the mounted files were presented), and only one reload of the certs is necessary. 
Based on my testing, if there's an update on a file, it only reports the update on that file (i.e. create/update); it doesn't trigger a second update on the directory as well.

Comment (tmshort) [reply internal/httputil/certpoolwatcher.go:63]: I added debug output to the cert watcher test for every event in the unit test:
```
=== RUN   TestCertPoolWatcher
    certpoolwatcher_test.go:72: Create cert file at "/tmp/cert-pool4285276756/test1.pem"
    certpoolwatcher_test.go:87: Create cert file at "/tmp/cert-pool4285276756/test2.pem"
Event: CREATE        "/tmp/cert-pool4285276756/test2.pem"
Event: WRITE         "/tmp/cert-pool4285276756/test2.pem"
--- PASS: TestCertPoolWatcher (1.11s)
```
So, there's an event for the create of the new PEM, and one for the write, but nothing on the directory itself. There are two events, and that would cause two reloads without the drain mechanism in place.

**Pull request merged**
